### DISCUSSION TITLE
Applying Dense()

### DISCUSSION THREAD

**Discussion by EngineerKhan** [google/flax#1667] [Q&A] [answered]
I want to learn `Dense()` (and other nn modules) by trying out very simple examples. But whenever I try to implement the [PyTorch counterpart ](https://pytorch.org/docs/stable/generated/torch.nn.Linear.html#torch.nn.Linear), it throws me the error:

>CallCompactUnboundModuleError: Can't call compact methods on unbound modules (https://flax.readthedocs.io/en/latest/flax.errors.html#flax.errors.CallCompactUnboundModuleError)

Here's my Flax translation:

```
m = nn.Dense(20, 30)
input = jax.random.uniform(jax.random.PRNGKey(0),(128, 20))
output = m(input)
print(output.size())
```
How can I fix it?

**Comment by matthias-wright** [comment, marked as answer]
Hi @EngineerKhan, 

PyTorch and Flax handle parameters and state very differently.
A PyTorch module holds all of its trainable parameters and other variables (such as the running_mean for BatchNorm).
Flax on the other hand takes a functional approach. When you initialize a Flax module, it returns all of its parameters and variables (see `params` below) instead of keeping track of them internally. This means that you have to pass them into the module along with your input whenever you want to compute the output (this is done with the `apply` method).
I changed your example, now it should work:

```python
import flax.linen as nn
import jax
  
rng = jax.random.PRNGKey(0)
init_rng, rng = jax.random.split(rng)

# create module
m = nn.Dense(20, 30)

# initialize params
x = jax.random.uniform(rng,(128, 20))
params = m.init(init_rng, x)

# compute output
output = m.apply(params, x)

print(output.shape)
```

For more info, check out [this](https://flax.readthedocs.io/en/latest/notebooks/flax_basics.html) page.

Hope that helps!

Edit: the `x` that I passed to the `init` function can be a dummy variable, e.g. `params = m.init(init_rng, jax.numpy.ones((1, 20)))`.

**Reply by EngineerKhan** [reply]
Thanks a lot. Is this `init` and `apply` similar to how we perform it in Haiku?

**Reply by EngineerKhan** [reply]
how and why**

**Reply by matthias-wright** [reply]
Yes
